cpp/tests/rolling/range_window_type_test.cpp

```diff
@@ -0,0 +1,1559 @@
+        // clang-format on
+      ascending_nulls_before{
+        {min, T{5}, T{5}, T{6}, T{7}, T{9}, T{9}, T{12}, T{13}, T{17}, T{22}, T{22}, max},
+        {false, false, false, false, true, true, true, true, true, true, true, true, true}},
+      ascending_nulls_after{
+        {min, T{5}, T{5}, T{6}, T{7}, T{9}, T{9}, T{12}, T{13}, T{17}, T{22}, T{22}, max},
+        {true, true, true, true, true, true, false, false, false, false, false, false, false}},
+      descending_no_nulls{
+        {max, T{22}, T{22}, T{17}, T{13}, T{12}, T{9}, T{9}, T{7}, T{6}, T{5}, T{5}, min}},
+      descending_nulls_before{
+        {max, T{22}, T{22}, T{17}, T{13}, T{12}, T{9}, T{9}, T{7}, T{6}, T{5}, T{5}, min},
+        {true, true, true, true, true, true, true, false, false, false, false, false, false}},
+      descending_nulls_after{
+        {max, T{22}, T{22}, T{17}, T{13}, T{12}, T{9}, T{9}, T{7}, T{6}, T{5}, T{5}, min},
+        {false, false, false, false, false, true, true, true, true, true, true, true, true}}
+  {
+  }
+
+  void run_test(cudf::column_view const& orderby,
+                cudf::order order,
+                cudf::null_order null_order,
+                cudf::range_window_type preceding,
+                cudf::range_window_type following,
+                cudf::column_view const& expect_preceding,
+                cudf::column_view const& expect_following)
+  {
+    auto result = cudf::make_range_windows(
+      cudf::table_view{}, orderby, order, null_order, preceding, following);
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(
+      std::get<0>(result)->view(), expect_preceding, cudf::test::debug_output_level::ALL_ERRORS);
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(
+      std::get<1>(result)->view(), expect_following, cudf::test::debug_output_level::ALL_ERRORS);
+  }
+};
+template <typename T>
+struct UngroupedIntegralRangeWindows : UngroupedBase<T> {};
+TYPED_TEST_SUITE(UngroupedIntegralRangeWindows, cudf::test::IntegralTypesNotBool);
+
+template <typename T>
+struct UngroupedSignedIntegralRangeWindows : UngroupedBase<T> {};
+using SignedIntegralTypes = cudf::test::Types<int8_t, int16_t, int32_t, int64_t>;
+TYPED_TEST_SUITE(UngroupedSignedIntegralRangeWindows, SignedIntegralTypes);
+
+TYPED_TEST(UngroupedIntegralRangeWindows, AscendingNoNullsZeroPrecedingFollowing)
+{
+  using T   = TypeParam;
+  auto prec = cudf::make_fixed_width_scalar<T>(0);
+  auto foll = cudf::make_fixed_width_scalar<T>(0);
+
+  this->run_test(this->ascending_no_nulls,
+                 cudf::order::ASCENDING,
+                 cudf::null_order::BEFORE,
+                 cudf::bounded_closed{*prec},
+                 cudf::bounded_closed{*foll},
+                 cudf::test::fixed_width_column_wrapper<cudf::size_type>{
+                   {1, 1, 2, 1, 1, 1, 2, 1, 1, 1, 1, 2, 1}},
+                 cudf::test::fixed_width_column_wrapper<cudf::size_type>{
+                   {0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0}});
```

Comment:
This is the part I was a little concerned about.  It is important that for `CURRENT_ROW` operations, `expected_following` should cover the last occurrence of the current row's value.  i.e. `expected_following[1] == 1`, `expected_following[2] == 2`, etc.

Thank you for covering this.